create astro error

### What version of `astro` are you using?

create-astro@3.1.7

### Are you using an SSR adapter? If so, which one?

none

### What package manager are you using?

pnpm

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

```
$ pnpm create astro
.../Library/pnpm/store/v3/tmp/dlx-42001  | +159 ++++++++++++++++
Packages are hard linked from the content-addressable store to the virtual store.
  Content-addressable store is at: /Users/magicdawn/Library/pnpm/store/v3
  Virtual store is at:             ../../Library/pnpm/store/v3/tmp/dlx-42001/node_modules/.pnpm
.../Library/pnpm/store/v3/tmp/dlx-42001  | Progress: resolved 159, reused 159, downloaded 0, added 159, done
undefined:1
<h1>404 Not Found</h1>
^

SyntaxError: Unexpected token < in JSON at position 0
    at JSON.parse (<anonymous>)
    at IncomingMessage.<anonymous> (file:///Users/magicdawn/Library/pnpm/store/v3/tmp/dlx-42001/node_modules/.pnpm/create-astro@3.1.7/node_modules/create-astro/dist/index.js:284:32)
    at IncomingMessage.emit (node:events:525:35)
    at endReadableNT (node:internal/streams/readable:1359:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)

Node.js v18.14.2
 ERROR  Command failed with exit code 1: create-astro

```

- create-astro@3.1.7/node_modules/create-astro/dist/index.js:284:32
- https://unpkg.com/browse/create-astro@3.1.7/dist/index.js

![image](https://github.com/withastro/astro/assets/4067115/acc3ca09-d952-42ce-8d7d-d32c4cd01c74)

```
$ pnpm config get registry
https://registry.npmmirror.com/

$ curl -I `pnpm config get registry`/astro/latest
HTTP/1.1 200 Connection established

HTTP/2 404
server: Tengine
content-type: text/html; charset=utf-8
content-length: 22
strict-transport-security: max-age=5184000
date: Sun, 11 Jun 2023 01:37:57 GMT
vary: Origin
x-frame-options: SAMEORIGIN
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
x-download-options: noopen
x-readtime: 0.804
ali-swift-global-savetime: 1686447477
via: cache44.l2cn3069[24,24,404-1280,M], cache42.l2cn3069[39,0], kunlun2.cn5595[59,58,404-1280,M], kunlun10.cn5595[60,0]
x-cache: MISS TCP_MISS dirn:-2:-2
x-swift-savetime: Sun, 11 Jun 2023 01:37:57 GMT
x-swift-cachetime: 1
x-swift-error: orig response 4XX error
timing-allow-origin: *
eagleid: 01c6049e16864474770991332e

$ curl -I `pnpm config get registry`astro/latest
HTTP/1.1 200 Connection established

HTTP/2 200
server: Tengine
content-type: application/json; charset=utf-8
content-length: 8239
strict-transport-security: max-age=5184000
date: Sun, 11 Jun 2023 01:37:40 GMT
vary: Origin, Accept, Accept-Encoding
request-id: 8dccb0f0-07f8-11ee-9757-8fb30df599d3
cache-control: public, max-age=300
x-frame-options: SAMEORIGIN
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
x-download-options: noopen
x-readtime: 42.56
via: cn5595.l1, kunlun15.cn5595, l2cn3069.l2, cache11.l2cn3069, npmmirror-x86-20220823002, cache11.l2cn3069[68,67,200-0,M], cache56.l2cn3069[69,0], kunlun15.cn5595[0,0,200-0,H], kunlun20.cn5595[1,0]
ali-swift-global-savetime: 1686447460
age: 21
x-cache: HIT TCP_MEM_HIT dirn:10:1135070938
x-swift-savetime: Sun, 11 Jun 2023 01:37:40 GMT
x-swift-cachetime: 300
timing-allow-origin: *
eagleid: 01c604a816864474811333530e
```

since this code uses `registry + /astro/latest`, a trim last `/` action is need for registry.
can not rely on registry to support `<registry>//:package-name` route.

I have edited the code, but PR need test, so I turned to Issues.
https://github.com/withastro/astro/compare/main...magicdawn:astro:patch-1


### Link to Minimal Reproducible Example

none

### Participation

- [ ] I am willing to submit a pull request for this issue.

(base) wenke@wenkedeMac-mini github % npm create astro@latest

undefined:1
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
^

SyntaxError: Unexpected token < in JSON at position 0
    at JSON.parse (<anonymous>)
    at IncomingMessage.<anonymous> (file:///Users/wenke/.npm/_npx/abdb4b598af046c4/node_modules/create-astro/dist/index.js:284:32)
    at IncomingMessage.emit (node:events:539:35)
    at endReadableNT (node:internal/streams/readable:1344:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)

Node.js v18.2.0
npm ERR! code 1
npm ERR! path /Users/wenke/github
npm ERR! command failed
npm ERR! command sh -c create-astro

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/wenke/.npm/_logs/2023-06-11T05_58_56_583Z-debug-0.log


mac os

--- a/.changeset/old-seahorses-fold.md
+++ b/.changeset/old-seahorses-fold.md
@@ -3,4 +3,4 @@
 'create-astro': patch
 ---
 
-Update registry logic to handle registry URLs ending with '/'
+Update registry logic, improving edge cases (redirects, UTF-8 sequences, registries ending with '/')

--- a/.changeset/old-seahorses-fold.md
+++ b/.changeset/old-seahorses-fold.md
@@ -2,4 +2,4 @@
 'create-astro': patch
 ---
 
-fix: getVersion Error when registry ends with '/'
+Update registry logic to handle registry URLs ending with '/'

--- a/.changeset/old-seahorses-fold.md
+++ b/.changeset/old-seahorses-fold.md
@@ -1,4 +1,5 @@
 ---
+'astro': patch
 'create-astro': patch
 ---
 

--- a/.changeset/old-seahorses-fold.md
+++ b/.changeset/old-seahorses-fold.md
@@ -1,0 +1,5 @@
+---
+'create-astro': major
+---
+
+fix: getVersion Error when registry ends with '/'

--- a/.changeset/old-seahorses-fold.md
+++ b/.changeset/old-seahorses-fold.md
@@ -1,5 +1,5 @@
 ---
-'create-astro': major
+'create-astro': patch
 ---
 
 fix: getVersion Error when registry ends with '/'

--- a/.changeset/old-seahorses-fold.md
+++ b/.changeset/old-seahorses-fold.md
@@ -3,4 +3,4 @@
 'create-astro': patch
 ---
 
-Update registry logic, improving edge cases (redirects, UTF-8 sequences, registries ending with '/')
+Update registry logic, improving edge cases (http support, redirects, registries ending with '/')

--- a/packages/astro/src/core/add/index.ts
+++ b/packages/astro/src/core/add/index.ts
@@ -79,7 +79,7 @@ async function getRegistry(): Promise<string> {
 	const packageManager = (await preferredPM(process.cwd()))?.name || 'npm';
 	try {
 		const { stdout } = await execa(packageManager, ['config', 'get', 'registry']);
-		return stdout || 'https://registry.npmjs.org';
+		return stdout?.trim()?.replace(/\/$/, '') || 'https://registry.npmjs.org';
 	} catch (e) {
 		return 'https://registry.npmjs.org';
 	}

--- a/packages/create-astro/package.json
+++ b/packages/create-astro/package.json
@@ -36,6 +36,7 @@
     "execa": "^6.1.0",
     "giget": "1.0.0",
     "mocha": "^9.2.2",
+    "node-fetch-native": "^1.2.0",
     "which-pm-runs": "^1.1.0"
   },
   "devDependencies": {

--- a/packages/create-astro/src/messages.ts
+++ b/packages/create-astro/src/messages.ts
@@ -81,7 +81,11 @@ let v: string;
 export const getVersion = () =>
 	new Promise<string>(async (resolve) => {
 		if (v) return resolve(v);
-		const registry = await getRegistry();
+		let registry = await getRegistry();
+		// if the registry ends with '/', trim it.
+		if (registry.lastIndexOf('/') === registry.length - 1) {
+			registry = registry.substring(0, registry.length - 1);
+		}
 		get(`${registry}/astro/latest`, (res) => {
 			let body = '';
 			res.on('data', (chunk) => (body += chunk));

--- a/packages/create-astro/src/messages.ts
+++ b/packages/create-astro/src/messages.ts
@@ -15,7 +15,7 @@ async function getRegistry(): Promise<string> {
 	const packageManager = detectPackageManager()?.name || 'npm';
 	try {
 		const { stdout } = await execa(packageManager, ['config', 'get', 'registry']);
-		return stdout || 'https://registry.npmjs.org';
+		return stdout?.trim()?.replace(/\/$/, '') || 'https://registry.npmjs.org';
 	} catch (e) {
 		return 'https://registry.npmjs.org';
 	}
@@ -82,10 +82,6 @@ export const getVersion = () =>
 	new Promise<string>(async (resolve) => {
 		if (v) return resolve(v);
 		let registry = await getRegistry();
-		// if the registry ends with '/', trim it.
-		if (registry.lastIndexOf('/') === registry.length - 1) {
-			registry = registry.substring(0, registry.length - 1);
-		}
 		get(`${registry}/astro/latest`, (res) => {
 			let body = '';
 			res.on('data', (chunk) => (body += chunk));

--- a/packages/create-astro/src/messages.ts
+++ b/packages/create-astro/src/messages.ts
@@ -3,7 +3,7 @@ import { color, label, say as houston, spinner as load } from '@astrojs/cli-kit'
 import { align, sleep } from '@astrojs/cli-kit/utils';
 import { execa } from 'execa';
 import { exec } from 'node:child_process';
-import { get } from 'node:https';
+import fetch from 'node-fetch-native';
 import stripAnsi from 'strip-ansi';
 import detectPackageManager from 'which-pm-runs';
 
@@ -82,15 +82,9 @@ export const getVersion = () =>
 	new Promise<string>(async (resolve) => {
 		if (v) return resolve(v);
 		let registry = await getRegistry();
-		get(`${registry}/astro/latest`, (res) => {
-			let body = '';
-			res.on('data', (chunk) => (body += chunk));
-			res.on('end', () => {
-				const { version } = JSON.parse(body);
-				v = version;
-				resolve(version);
-			});
-		});
+		const { version } = await fetch(`${registry}/astro/latest`, { redirect: 'follow' }).then(res => res.json());
+		v = version;
+		resolve(version);
 	});
 
 export const log = (message: string) => stdout.write(message + '\n');

